internal/component/database_observability/mysql/collector/query_sample.go

```diff
@@ -140,56 +142,53 @@ func (c *QuerySample) Stop() {
 	c.cancel()
 }
 
-// setLastSampleSeenTimestamp queries the database for the latest sample seen timestamp so that upon startup we don't collect
+// initializeBookmark queries the database for the uptime "modulo" overflows so that upon startup we don't collect
 // samples that may have been collected previously.
-func (c *QuerySample) setLastSampleSeenTimestamp(ctx context.Context) error {
-	rs, err := c.dbConnection.QueryContext(ctx, selectLatestTimerEnd)
+func (c *QuerySample) initializeBookmark(ctx context.Context) error {
+	rs := c.dbConnection.QueryRowContext(ctx, selectUptime)
+
+	var uptime float64
+	err := rs.Scan(&uptime)
 	if err != nil {
-		return fmt.Errorf("failed to fetch last sample seen timestamp: %w", err)
+		return fmt.Errorf("failed to scan uptime: %w", err)
 	}
-	defer rs.Close()
 
-	for rs.Next() {
-		var ts sql.NullFloat64
-		err := rs.Scan(&ts)
-		if err != nil {
-			return fmt.Errorf("failed to scan row: %w", err)
-		}
-		if !ts.Valid {
-			return fmt.Errorf("no valid timestamp found: %w", err)
-		}
-		c.lastSampleSeenTimestamp = ts.Float64
-	}
-	if err := rs.Err(); err != nil {
-		return fmt.Errorf("failed to iterate rows: %w", err)
-	}
+	c.lastUptime = uptime
+	c.timerBookmark = uptimeModuloOverflows(uptime)
 	return nil
 }
 
 func (c *QuerySample) fetchQuerySamples(ctx context.Context) error {
-	rs, err := c.dbConnection.QueryContext(ctx, selectQuerySamples, c.lastSampleSeenTimestamp)
+	timeRow := c.dbConnection.QueryRowContext(ctx, selectNowAndUptime)
+
+	var now, uptime float64
+	if err := timeRow.Scan(&now, &uptime); err != nil {
+		return fmt.Errorf("failed to scan now and uptime info: %w", err)
+	}
+
+	timerClause, limit := c.determineTimerClauseAndLimit(uptime)
+	rs, err := c.dbConnection.QueryContext(ctx, selectQuerySamples+timerClause, c.timerBookmark, limit)
 	if err != nil {
-		level.Error(c.logger).Log("msg", "failed to fetch history table samples", "err", err)
-		return err
+		return fmt.Errorf("failed to fetch query samples: %w", err)
 	}
 	defer rs.Close()
 
+	// set the new bookmarks
+	c.timerBookmark = limit
+	c.lastUptime = uptime
+
 	for rs.Next() {
 		row := struct {
-			// system times
-			NowSeconds    uint64
-			UptimeSeconds uint64
-
 			// sample metadata
 			Schema     sql.NullString
 			Digest     sql.NullString
 			DigestText sql.NullString
 
 			// sample time
-			TimerStartPicoseconds  sql.NullFloat64
 			TimerEndPicoseconds    sql.NullFloat64
-			ElapsedTimePicoseconds sql.NullInt64
-			CPUTime                uint64
+			TimestampMilliseconds  uint64
+			ElapsedTimePicoseconds sql.NullFloat64
+			CPUTime                float64
```

Comment:
Is `CPUTime` simply going to be zero if the `setup_consumer` is not enabled? Or could this be nullable?

Reply:
It will be zero